In [17]:
import pandas as pd
import math
import numpy as np
import chardet as cde
import loc 

In [7]:
math.factorial(5)

120

In [4]:
s = pd.Series([1,3,5, np.nan, 6 ,8 ])
print (s)

0    1.0
1    3.0
2    5.0
3    NaN
4    6.0
5    8.0
dtype: float64


In [18]:
# Lendo os bytes do arquivo com chardet-cde para identificar qual enconding/utf do mesmo
with open  ("Bases/amazon.csv","rb") as f: 
    result = cde.detect(f.read(10000))
print(result["encoding"])

ISO-8859-1


In [19]:
# Estilo padrão para ler arquivos no panda, porém pode ter conflito de utf-8 ]
# df = pd.read_csv ("Bases/amazon.csv")

#Após leitura do arquivo estamos utilizando o chardet - cde para armazenar o resultado na variavel code_utf para converter o encoding automaticamente.
#Estamos utilizando tbm o separador , e o header = none para realizar o tratamento da base onde desconsiderados a primeira linha como 'coluna' = lendo como texto
code_utf = result["encoding"]
df = pd.read_csv ("Bases/amazon.csv", encoding=code_utf , sep = ',', header=None)


In [20]:
# Processamento para corrigir as colunas afinal um dos problemas é a virgula na separação
df = df[0].str.split(',', expand=True)

#  Remover aspas extras dos nomes das colunas
df.columns = df.iloc[0].str.replace('"','')
df = df[1:] # Remove a linha de cabeçalho original

# Reseta índice
df = df.reset_index(drop=True)

In [23]:
# Atribuindo nomes corretos para as colunas
df.columns = ['year','state', 'month', 'number', 'date']

# Converter tipos de dados
df['year'] = df['year'].astype(int)
df['number'] = pd.to_numeric(df['number'], errors='coerce')

In [25]:
df.shape
print(f"Temos {df.shape[0]} linhas e {df.shape[1]} colunas.")

Temos 6454 linhas e 5 colunas.


1) Importe os dados e verifique as primeiras linhas do dataset. Verifique também as últimas linhas do dataset.

In [26]:
df.info

<bound method DataFrame.info of       year        state       month  number        date
0     1998       "Acre"   "Janeiro"     0.0  1998-01-01
1     1999       "Acre"   "Janeiro"     0.0  1999-01-01
2     2000       "Acre"   "Janeiro"     0.0  2000-01-01
3     2001       "Acre"   "Janeiro"     0.0  2001-01-01
4     2002       "Acre"   "Janeiro"     0.0  2002-01-01
...    ...          ...         ...     ...         ...
6449  2012  "Tocantins"  "Dezembro"   128.0  2012-01-01
6450  2013  "Tocantins"  "Dezembro"    85.0  2013-01-01
6451  2014  "Tocantins"  "Dezembro"   223.0  2014-01-01
6452  2015  "Tocantins"  "Dezembro"   373.0  2015-01-01
6453  2016  "Tocantins"  "Dezembro"   119.0  2016-01-01

[6454 rows x 5 columns]>

In [24]:
df.head(5)

,year,state,month,number,date
0,1998,"""Acre""","""Janeiro""",0.0,1998-01-01
1,1999,"""Acre""","""Janeiro""",0.0,1999-01-01
2,2000,"""Acre""","""Janeiro""",0.0,2000-01-01
3,2001,"""Acre""","""Janeiro""",0.0,2001-01-01
4,2002,"""Acre""","""Janeiro""",0.0,2002-01-01


In [11]:
#Criando lista para tratar as linhas da coluna state e month
limpar_aspas = ['state' , 'month']

#Removendo aspas das colunas
df[limpar_aspas] = df[limpar_aspas].apply(lambda x: x.str.replace ('"', ''))

In [ ]:
# Mostrando os 5 primeiros valores
df.head(5)


,year,state,month,number,date
0,1998,Acre,Janeiro,0.0,1998-01-01
1,1999,Acre,Janeiro,0.0,1999-01-01
2,2000,Acre,Janeiro,0.0,2000-01-01
3,2001,Acre,Janeiro,0.0,2001-01-01
4,2002,Acre,Janeiro,0.0,2002-01-01


In [44]:
# capturando linhas onde o estado é Bahia e possui números maiores que 0
print(df.loc[(df['state'].str.lower() == "bahia") & (df['number'] > 0)])

0     year  state     month   number        date
958   1999  Bahia   Janeiro  114.000  1999-01-01
959   2000  Bahia   Janeiro   31.000  2000-01-01
960   2001  Bahia   Janeiro   24.000  2001-01-01
961   2002  Bahia   Janeiro  125.000  2002-01-01
962   2003  Bahia   Janeiro  439.000  2003-01-01
...    ...    ...       ...      ...         ...
1191  2012  Bahia  Dezembro  477.000  2012-01-01
1192  2013  Bahia  Dezembro  131.000  2013-01-01
1193  2014  Bahia  Dezembro  144.000  2014-01-01
1194  2015  Bahia  Dezembro    1.408  2015-01-01
1195  2016  Bahia  Dezembro  164.000  2016-01-01

[234 rows x 5 columns]


In [ ]:
# Mostrando possíveis valores nulos, funciona apenas para números
df.describe()

,year,number
count,6454.000000,6454.000000
mean,2007.461729,108.293163
std,5.746654,190.812242
min,1998.000000,0.000000
25%,2002.000000,3.000000
50%,2007.000000,24.000000
75%,2012.000000,113.000000
max,2017.000000,998.000000


In [ ]:
# Mostrando colunas state e month, para ver se possui alguma nula, funciona para str
df[['state', 'month']].value_counts

<bound method DataFrame.value_counts of 0         state     month
0          Acre   Janeiro
1          Acre   Janeiro
2          Acre   Janeiro
3          Acre   Janeiro
4          Acre   Janeiro
...         ...       ...
6449  Tocantins  Dezembro
6450  Tocantins  Dezembro
6451  Tocantins  Dezembro
6452  Tocantins  Dezembro
6453  Tocantins  Dezembro

[6454 rows x 2 columns]>

In [48]:
# Utilizando group by para capturar os numeros medios de todos os estados
df.groupby("state")["number"].mean()

state
Acre                 77.255356
Alagoas              19.350000
Amapa                91.345506
Amazonas            128.243218
Bahia               187.222703
Ceara               127.314071
Distrito Federal     14.899582
Espirito Santo       27.389121
Goias               157.721841
Maranhao            105.142808
Mato Grosso         201.351523
Minas Gerais        156.800243
Paraiba             109.698573
Pará                102.561272
Pernambuco          102.502092
Piau                158.174674
Rio                  62.985865
Rondonia             84.876272
Roraima             102.029598
Santa Catarina      101.924067
Sao Paulo           213.896226
Sergipe              13.543933
Tocantins           141.037176
Name: number, dtype: float64

2) Verifique se existe algum valor missing no dataset.

3) Quanto de memória o dataset está consumindo?

4) Gere as estatísticas descritivas das colunas numéricas do dataset. Qual a mediana da coluna BMI? E a média? Qual desses valores é maior? O que isso significa? Caso não saiba o que isso significa, passe os 2 valores ao chatGPT e pergunte a ele o que significa a média ser maior ou menor que a mediana.

5) Qual a proporção de fumantes no nosso dataset? E de não fumantes?

6) Existe alguma diferença na cobrança para fumantes e não fumantes? Se sim, ao que você imagina que isso se deve?

7) Qual a média de BMI para fumantes? Ela é muito diferente da observada para fumantes? Se sim, por qual motivo você acha que isso ocorre?

8) Existe alguma diferença de cobrança para quem tem filhos? Identifique se ter filhos parece acarretar numa cobrança maior olhando a média de charges para quem tem 0 filhos e a média para quem tem 1 ou mais filhos.

9) Para o homem, o BMI de 20 a 24,9 é considerado normal. Acima disso, é considerado acima do peso, abaixo é considerado abaixo do peso. Crie uma nova coluna, chamada "bmi_male" que receberá o texto "abaixo do normal" para BMI abaixo de 20, "normal" para BMI de 20 a 24,9 e "acima do normal" para BMI acima de 24,9.

10) Se uma homem é fumante e ainda tem um BMI acima do considerado normal, é possível que isso seja um indicador negativo para a seguradora. Quantos homens estão nessa situação?

Utilizando o dataset de incêndios florestais no Brasil, que você pode baixar clicando aqui, responda:

11) Carregue o dataset e crie uma mensagem ao leitor do notebook falando quantas linhas e colunas possui o DataFrame.

12) Utilize loc e iloc para exibir da linha 100 até a linha 120 da segunda, terceira e quarta coluna. Qual a diferença das duas sintaxes feitas?

13) Qual o total de incêndios que ocorreram no Brasil no ano de 2010? E se considerarmos somente o estado do Mato Grosso do Sul?

14) Qual a média de incêndios por mês no estado do Acre? E o desvio padrão?

15) Qual estado possui a maior média de incêndios por mês? E qual possui a menor?

16) Crie uma coluna chamada regiao_sudeste e marque com 1 para quando a observação pertencer a um destes estados e 0, caso contrário. Faça o mesmo para a região centro-oeste. Qual a diferença no total de incêndios dessas regiões? Qual região possui mais incêndios? Consulte a internet para entender por qual motivo uma das regiões possui mais incêndio que a outra.